experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,694 @@
+  end
+  R = base_ring(parent(f[1]))
+  for a in f
+    R == base_ring(parent(a)) || error("fractions do not belong to the same ring")
+  end
+  R == base_ring(OO(X)) || error("fractions do not belong to the base ring of the scheme")
+  a = numerator.(f)
+  b = denominator.(f)
+  W = localized_ring(OO(X))
+  I = ideal(W, one(W))
+  for p in f
+    I = intersect(quotient(ideal(W, denominator(p)) + localized_modulus(OO(X)), ideal(W, numerator(p))), I)
+  end
+  U = SpecOpen(X, I)
+  return U, [StructureSheafElem(U, [OO(V)(a) for V in patches(U)]) for a in f]
+end
+
+@Markdown.doc """
+    SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2}
+
+Morphisms ``f : U → V`` of open sets ``U ⊂ X`` and ``V ⊂ Y`` of affine schemes.
+These are stored as morphisms ``fᵢ: Uᵢ→ Y`` on the affine patches 
+``Uᵢ`` of ``U``.
+"""
+mutable struct SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2}
+  domain::SpecOpen{BRT, BRET, RT, RET, MST1}
+  codomain::SpecOpen{BRT, BRET, RT, RET, MST2}
+  maps_on_patches::Vector{SpecMor{BRT, BRET, RT, RET, MST1, MST2}}
+
+  # fields used for caching
+  inverse::SpecOpenMor{BRT, BRET, RT, RET, MST2, MST1}
+
+  function SpecOpenMor(
+      U::SpecOpen{BRT, BRET, RT, RET, MST1},
+      V::SpecOpen{BRT, BRET, RT, RET, MST2},
+      f::Vector{SpecMor{BRT, BRET, RT, RET, MST1, MST2}};
+      check::Bool=true
+    ) where {BRT, BRET, RT, RET, MST1, MST2}
+    Y = ambient(V)
+    n = length(f)
+    n == length(patches(U)) || error("number of patches does not coincide with the number of maps")
+    if check
+      for i in 1:n
+	domain(f[i]) == patches(U)[i] || error("domain of definition of the map does not coincide with the patch")
+	codomain(f[i]) == Y || error("codomain is not compatible")
+      end
+      for i in 1:n-1
+	for j in i+1:n
+	  A = intersect(domain(f[i]), domain(f[j]))
+	  restrict(f[i], A, Y) == restrict(f[j], A, Y) || error("maps don't glue")
+	end
+      end
+      I = ideal(localized_ring(OO(Y)), gens(V))
+      for g in f
+	one(localized_ring(OO(domain(g)))) in pullback(g)(I) + localized_modulus(OO(domain(g))) || error("image is not contained in the codomain")
+      end
+    end
+    return new{BRT, BRET, RT, RET, MST1, MST2}(U, V, f)
+  end
+end
+
+domain(f::SpecOpenMor) = f.domain
+codomain(f::SpecOpenMor) = f.codomain
+maps_on_patches(f::SpecOpenMor) = f.maps_on_patches
+getindex(f::SpecOpenMor, i::Int) = maps_on_patches(f)[i]
+
+function SpecOpenMor(U::SpecOpenType, V::SpecOpenType, f::Vector) where {SpecOpenType<:SpecOpen}
+  Y = ambient(V)
+  return SpecOpenMor(U, V, [SpecMor(W, Y, f) for W in patches(U)]) 
+end
+
+function inclusion_morphism(U::SpecOpenType, V::SpecOpenType) where {SpecOpenType<:SpecOpen}
+  X = ambient(U)
+  ambient(V) == X || error("method not implemented")
+  return SpecOpenMor(U, V, gens(base_ring(OO(X))))
+end
+
+
+@Markdown.doc """
+    compose(f::T, g::T) where {T<:SpecOpenMor}
+
+computes the composition of two morphisms 
+
```

Comment:
does this render acceptably in the documentation?

Reply:
I wanted to give it a try. It should print as code, so indentation should be kept. We'll see.